src/one/nio/ws/WebSocketSession.java

```diff
@@ -0,0 +1,211 @@
+        if (message instanceof PingMessage) {
+            server.handleMessage(session, (PingMessage) message);
+        } else if (message instanceof PongMessage) {
+            server.handleMessage(session, (PongMessage) message);
+        } else if (message instanceof TextMessage) {
+            server.handleMessage(session, (TextMessage) message);
+        } else if (message instanceof BinaryMessage) {
+            server.handleMessage(session, (BinaryMessage) message);
+        } else if (message instanceof CloseMessage) {
+            server.handleMessage(session, (CloseMessage) message);
+        }
+    }
+
+    @Override
+    public void handleException(Throwable e) {
+        log.error(e);
+        if (e instanceof WebSocketException) {
+            close(((WebSocketException) e).code());
+            return;
+        }
+        super.handleException(e);
+    }
+
+    public void close(short code) {
+        try {
+            sendMessage(new CloseMessage(code));
+        } catch (Exception e) {
+            log.warn("error while sending closing frame", e);
+        } finally {
+            close();
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            extensions.forEach(Extension::close);
+        } finally {
+            super.close();
+        }
+    }
+
+    protected void sendHandshakeResponse(Request request) throws IOException {
+        try {
+            final Response response = createResponse(request);
+            processExtensions(request, response);
+            processProtocol(request, response);
+            reader = new MessageReader(this, extensions, config.maxFramePayloadLength, config.maxMessagePayloadLength);
+            writer = new MessageWriter(this, extensions);
+            sendResponse(response);
+        } catch (HandshakeException e) {
+            sendError(Response.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    protected Response createResponse(Request request) {
+        try {
+            Response response = new Response(Response.SWITCHING_PROTOCOLS, Response.EMPTY);
+            response.addHeader("Upgrade: websocket");
+            response.addHeader("Connection: Upgrade");
+            response.addHeader(WebSocketHeaders.createAcceptHeader(request));
+            return response;
+        } catch (VersionException e) {
+            Response response = new Response("426 Upgrade Required", Response.EMPTY);
+            response.addHeader(WebSocketHeaders.createVersionHeader(13));
+            return response;
+        }
+    }
+
+    protected void processProtocol(Request request, Response response) {
+        final String protocols = request.getHeader(WebSocketHeaders.PROTOCOL);
+        if (protocols != null) {
+            for (String protocol : protocols.split(",")) {
+                if (config.isSupportedProtocol(protocol)) {
+                    response.addHeader(WebSocketHeaders.PROTOCOL + protocol);
```

Comment:
PROTOCOL header value from `request` is transformed into **multiple** headers in `response` -- is that the desired behavior?

Reply:
there is a `break` on next line, so only first supported protocol will be taken